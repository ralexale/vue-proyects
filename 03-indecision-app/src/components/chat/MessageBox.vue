<template>
  <div class="bg-white p-4 flex items-center">
    <input
      type="text"
      v-model="message"
      @keypress.enter="sendMessage"
      placeholder="Type your message..."
      class="flex-1 border rounded-full px-4 py-2 focus:outline-none"
    />
    <button
      @click="sendMessage"
      class="bg-blue-500 text-white rounded-full p-2 ml-2 hover:bg-blue-600 focus:outline-none"
    >
      <img src="/icons/chat/sendIcon.svg" alt="send icon" />
    </button>
  </div>
</template>

<script lang="ts" setup>
import { ref } from 'vue';

const message = ref<string>('');

const emit = defineEmits<{
  sendMessage: [message: string];
}>();

const sendMessage = async () => {
  if (!message.value) return;
  emit('sendMessage', message.value);
  message.value = '';
};
</script>
