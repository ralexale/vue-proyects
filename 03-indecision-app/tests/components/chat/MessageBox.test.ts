import { mount } from '@vue/test-utils';
import MessageBox from '@/components/chat/MessageBox.vue';

describe('<MessageBox />', () => {
  const wrapper = mount(MessageBox);
  const input = wrapper.find('input[type="text"]');
  const button = wrapper.find('button');

  test('should render inputs and buttons correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
    expect(input.exists()).toBe(true);
    expect(button.exists()).toBe(true);
    expect(wrapper.find('button img').exists()).toBe(true);
  });

  test('emits sendMessage event when button is clicked and have a message', async () => {
    // preparación
    const message = 'hola mundo';

    await input.setValue(message);
    await button.trigger('click');

    expect(wrapper.emitted('sendMessage')).toBeTruthy();
    expect(wrapper.emitted('sendMessage')?.[0]).toEqual([message]);
    expect((wrapper.vm as any).message).toBe('');
  });

  test('emits sendMessage event when enter key is pressed and have a message', async () => {
    // preparación
    const message = 'hola mundo';

    await input.setValue(message);
    await input.trigger('keypress.enter');

    expect(wrapper.emitted('sendMessage')).toBeTruthy();
    expect(wrapper.emitted('sendMessage')?.[0]).toEqual([message]);
    expect((wrapper.vm as any).message).toBe('');
  });

  test('no emits sendMessage event when enter key and button is pressed if message is empty', async () => {
    const wrapper = mount(MessageBox);

    await input.trigger('keypress.enter');
    await button.trigger('click');
    
    expect(wrapper.emitted('sendMessage')).toBeFalsy();
  });
});
