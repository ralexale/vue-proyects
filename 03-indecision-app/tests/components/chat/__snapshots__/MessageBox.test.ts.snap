// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<MessageBox /> > should render inputs and buttons correctly 1`] = `"<div class="bg-white p-4 flex items-center"><input type="text" placeholder="Type your message..." class="flex-1 border rounded-full px-4 py-2 focus:outline-none"><button class="bg-blue-500 text-white rounded-full p-2 ml-2 hover:bg-blue-600 focus:outline-none"><img src="/icons/chat/sendIcon.svg" alt="send icon"></button></div>"`;
